[clang][CodeGen] Emit `llvm.tbaa.errno` metadata during module creation

antoniofrighetto opened the pull request:
Let Clang emit `llvm.tbaa.errno` metadata in order to let LLVM carry out optimizations around errno-writing libcalls, as long as it is proved the involved memory location does not alias `errno`.

Previous discussion: https://discourse.llvm.org/t/rfc-modelling-errno-memory-effects/82972.

antoniofrighetto commented:
@nikic Mind glancing over the draft quickly? !llvm.errno.tbaa is being attached to individual load/store accessing errno, although I just realized that this information is already embedded in TBAA, so IIUC !llvm.errno.tbaa should just be a module-level list of int-compatible TBAA nodes. Is my latest understanding correct? That would make it much simpler than I thought. Sorry if I’m missing something!

nikic commented:
> so IIUC !llvm.errno.tbaa should just be a module-level list of int-compatible TBAA nodes.

Yes, exactly. It should be a module level MD node, not on individual instructions.

nikic commented:
> > > so IIUC !llvm.errno.tbaa should just be a module-level list of int-compatible TBAA nodes.
> > 
> > 
> > Yes, exactly. It should be a module level MD node, not on individual instructions.
> 
> Thank you, I'm updating this. I've been contemplating this, and thought originally we needed to extend TBAA (like aliasErrno as per tmp commit, provided it is correct), but TBAA seems to take care of this already (as clang propagates int TBAA correctly to the call), so I think this shouldn't be needed.

The fact that clang currently places TBAA on FP libcalls is a hack that we want to remove. It doesn't work for libcalls that also access other memory.

antoniofrighetto commented:
> > > > so IIUC !llvm.errno.tbaa should just be a module-level list of int-compatible TBAA nodes.
> > > 
> > > 
> > > Yes, exactly. It should be a module level MD node, not on individual instructions.
> > 
> > 
> > Thank you, I'm updating this. I've been contemplating this, and thought originally we needed to extend TBAA (like aliasErrno as per tmp commit, provided it is correct), but TBAA seems to take care of this already (as clang propagates int TBAA correctly to the call), so I think this shouldn't be needed.
> 
> The fact that clang currently places TBAA on FP libcalls is a hack that we want to remove. It doesn't work for libcalls that also access other memory.

Right, that makes sense! Removing this from draft, hopefully it's heading in the correct direction.

antoniofrighetto commented:
Rebased over 38e8dff84ba1e2cf0260712d21dd429d74471d08.

antoniofrighetto commented:
Kind ping.

antoniofrighetto commented:
> Sorry, can you back up and explain the root problem you're trying to solve? You have a platform where we're emitting accesses to `errno` as if it weren't an `int`? What does it look like in source?

Oh, sorry for not providing context earlier. The reasoning behind is to allow certain optimizations involving errno-writing libcalls to occur, if we are able to prove the involved memory location does not alias errno (e.g., because we are accessing a float-typed pointer). IIUC, in standard C, errno is an int, so all legitimate accesses must use int-compatible types (or otherwise violating strict aliasing). Hence, we should be able to leverage TBAA and prove not aliasing with errno, if we do not alias for any nodes belonging to `!llvm.errno.tbaa` (which is presumably gathering all int-based nodes; this should require confirmation).

Original miscompilation issue: https://github.com/llvm/llvm-project/issues/114772, previous discussion at: https://discourse.llvm.org/t/rfc-modelling-errno-memory-effects/82972.

rjmccall commented:
> Oh, sorry for not providing context earlier. The reasoning behind is to allow certain optimizations involving errno-writing libcalls (marked as `memory(errnomem: write)`) to occur, if we are able to prove the involved memory location does not alias errno (e.g., because we are accessing a float-typed pointer). IIUC, in standard C, errno is an int, so all legitimate accesses must use int-compatible types (or otherwise violating strict aliasing). Hence, we should be able to leverage TBAA and prove not aliasing with errno, if we do not alias for any nodes belonging to `!llvm.errno.tbaa` (which is presumably gathering all int-based nodes; this should require confirmation).

I see. So the idea here is *not* that we're changing the TBAA nodes used for any actual accesses. Instead, we're using a named global metadata to specify the TBAA node that all `errno` accesses would use so that LLVM can do alias analysis around these libcalls (presumably to allow load/store optimization across them). That makes a lot of sense, thanks for the clarification. Please update your commit messages and PR title and description; I think if you re-read them, you're see that they sound as if you're trying to change how accesses to `errno` are annotated. In the meantime, I think I have the context to review the PR now.

rjmccall commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -1467,6 +1468,17 @@ void CodeGenModule::Release() {
   }
+
+  // Emit `!llvm.errno.tbaa`, a module-level metadata that specifies the TBAA
+  // for an integer access.
```
```suggestion
  // for an int access. This allows LLVM to reason about what memory can
  // be accessed by certain library calls that only touch errno.
```

rjmccall commented on clang/test/C/C11/n1285_1.c:
```diff
@@ -32,9 +32,9 @@ struct X f(void);
-// C11-O2-NEXT:    [[TMP1:%.*]] = load i32, ptr [[TMP0]], align 4, !tbaa [[TBAA7:![0-9]+]]
+// C11-O2-NEXT:    store ptr [[ARRAYDECAY]], ptr [[P]], align 8, !tbaa [[TBAA6:![0-9]+]]
+// C11-O2-NEXT:    [[TMP0:%.*]] = load ptr, ptr [[P]], align 8, !tbaa [[TBAA6]]
+// C11-O2-NEXT:    [[TMP1:%.*]] = load i32, ptr [[TMP0]], align 4, !tbaa [[TBAA2:![0-9]+]]
```
Since you're renaming these anyway, could you rename them to something semantic like `INTPTR_TBAA` and `INT_TBAA`? The number doesn't really mean anything to test readers, and when there are TBAA changes in the future, we want to just be able to change where `INTPTR_TBAA` is matched in the metadata section rather than having to renumber all the variables in the tests every time.

(Same for the other tests)

efriedma-quic replied:
The check lines here are autogenerated; if we edit anything by hand, we'll need to mark it as a test that can't be automatically regenerated.  Which annoying not only right now, but in the future if the test ever needs to be updated.

It might be possible to make update_cc_test_checks.py choose better names for TBAA checks; in theory, it could take the type name from the metadata, and use it to choose the name of the tbaa variable.

antoniofrighetto replied:
An attempt to address this at: https://github.com/llvm/llvm-project/pull/147670.

rjmccall reviewed (changes requested):
Functionality LGTM. I have a somewhat annoying request for the tests, but it should make things a lot easier for the next person.

nikic reviewed (commented):
* The metadata should be documented in LangRef.
* We should test linking of two modules with the metadata. Both the case where they match and where they don't (e.g. mixing C and C++). IRLinker::linkNamedMDNodes should be the relevant code. (I do wonder whether we should instead be making this a module flag with AppendUnique behavior. I'm not sure clear on what the tradeoff is between a module flag and separate named metadata.)

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -970,6 +971,19 @@ ModRefInfo BasicAAResult::getModRefInfo(const CallBase *Call,
+    bool IsLocSizeUnknown = Loc.Size == MemoryLocation::UnknownSize;
+    if (ErrnoMR == ModRefInfo::Mod && !isa<AllocaInst>(Object) &&
+        (IsLocSizeUnknown ||
+         (!IsLocSizeUnknown && Loc.Size.getValue() <= sizeof(int))) &&
```
Should not use `sizeof(int)` here, which is a host property. We should query the `int` size from TLI instead.

nikic commented on llvm/lib/Analysis/BasicAliasAnalysis.cpp:
```diff
@@ -970,6 +971,19 @@ ModRefInfo BasicAAResult::getModRefInfo(const CallBase *Call,
+    if (ErrnoMR == ModRefInfo::Mod && !isa<AllocaInst>(Object) &&
+        (IsLocSizeUnknown ||
+         (!IsLocSizeUnknown && Loc.Size.getValue() <= sizeof(int))) &&
+        AAQI.AAR.aliasErrno(Loc, Call->getModule()) != AliasResult::NoAlias)
```
Also, possibly these extra checks should be in BasicAA::aliasErrno instead?

nikic commented:
It would probably be best to split up the LLVM and the Clang changes.

antoniofrighetto commented on llvm/test/Transforms/InstCombine/may-alias-errno.ll:
```diff
@@ -0,0 +1,90 @@
+declare float @sinf(float noundef) memory(errnomem: write)
+declare void @escape(ptr %p)
+
+!llvm.errno.tbaa = !{!0}
```
Checks and test added, thanks.

antoniofrighetto commented:
> It would probably be best to split up the LLVM and the Clang changes.

Thank you, moved AA changes separately, addressed reviews. Updated this PR as well, pending an update on tests due to https://github.com/llvm/llvm-project/pull/147670.

> (I do wonder whether we should instead be making this a module flag with AppendUnique behavior. I'm not sure clear on what the tradeoff is between a module flag and separate named metadata.)

Looks like IRLinker::linkNamedMDNodes already disallows duplicated nodes, both are map lookups, thus took the choice of favouring a named metadata here (IR less verbose, getModuleFlag() would come with further checks in TBAA::aliasErrno; if I haven't missed anything).

nikic reviewed (commented):
This looks fine, but is there any test that actually checks for the presence of errno.tbaa? It looks like all of these are just updates due to numbering changes.

antoniofrighetto commented:
A kind ping for @rjmccall. As previously were requested changes, any other concern with this? (Only the Clang-side change has been left here.)

nikic reviewed (approved):
LGTM from my side.

rjmccall reviewed (commented):
To be honest, I was hoping the TBAA test-generation changes would leave us in a position where we *didn't* have a bunch of spurious changes to tests just because something unrelated perturbed the metadata order. It's great for readability that the metadata match variable includes the name of the type now, but the fact that it still includes a number means that this PR still has to have a bunch of changes renaming e.g. `CHAR_TBAA2` to `CHAR_TBAA6`.

I signed off on the functionality a long time ago, so if you want to commit, I'm not trying to block it.

antoniofrighetto commented:
> To be honest, I was hoping the TBAA test-generation changes would leave us in a position where we _didn't_ have a bunch of spurious changes to tests just because something unrelated perturbed the metadata order. It's great for readability that the metadata match variable includes the name of the type now, but the fact that it still includes a number means that this PR still has to have a bunch of changes renaming e.g. `CHAR_TBAA2` to `CHAR_TBAA6`.

If you are fine with that, I'd note this as a TODO in update_test_checks, and would be happy to do this in the future.

rjmccall commented:
Yes, I'm fine with you merging as-is.

nikic commented:
Reverse ping

antoniofrighetto committed:
[clang][CodeGen] Emit `llvm.tbaa.errno` metadata during module creation

Let Clang emit `llvm.tbaa.errno` metadata in order to let LLVM
carry out optimizations around errno-writing libcalls to, as
long as it is proved the involved memory location does not alias
`errno`.

Previous discussion: https://discourse.llvm.org/t/rfc-modelling-errno-memory-effects/82972.
